ESC Webhooks Launch Blog (#12767)
pulumi/docs: content/blog/esc-webhooks-launch/index.md

---
title: "Introducing Webhooks for Pulumi ESC"
date: 2024-09-13T00:00:00-03:00
draft: false
meta_desc: "Pulumi ESC Webhooks enable you to automate workflows, trigger actions, and streamline the management of your infrastructure and applications"
meta_image: "meta.png"
authors:
  - iaroslav-titov
  - arun-loganathan
tags:
  - esc
  - secrets
  - features
---

Managing secrets and configurations across multiple environments and teams can be a complex juggling act for development teams. [Pulumi ESC](/docs/esc/), our developer-friendly secrets management product, simplifies this process and accelerates your development cycle. Today, we're excited to make this process even easier with the launch of Webhooks for Pulumi ESC, a powerful new feature that allows you to send a custom trigger to any URL endpoint, enabling you to automate workflows and keep your infrastructure and applications up to date. Pulumi ESC Webhooks also integrates seamlessly with popular communication platforms like Slack and Microsoft Teams.

<!--more-->

> With Pulumi ESC Webhooks, you can now receive real-time notifications and trigger automated actions based on specific events within your Pulumi ESC [Environments](/docs/esc/environments/).  For example, you can set up a Webhook to notify your team’s Slack channel whenever a new secret is added to an environment, or automatically trigger a redeployment pipeline when configurations are changed. Webhooks can be triggered by changes to [version tags](/docs/esc/environments/#tagging-versions) and [environment tags](/blog/esc-projects-environment-tags-launch/#introducing-environment-tags), and can be configured per environment or at the organization level for all environments.

Reviewer: There is no webhook "when new secret is added to the environment'. I suggest we change this to "a new revision was created" or "a new tag is added to an environment revision"

Reviewer: nit: What's `redeployment`? I think just `deployment` gets the point across

Author: I'll update deployment...

re. "when new revision is created" - it is intentional. For folks who don't know revision or don't care and use the latest - just want to make it generic.

> Take your automation further by creating Webhooks that trigger [Pulumi Deployments](/docs/pulumi-cloud/deployments/), automatically updating your infrastructure in response to configuration changes in Pulumi ESC—without writing custom code and all within the Pulumi Cloud Console.

Reviewer: Since there's a whole blurb about it, is it worth mentioning above?

Author: What would you suggest we add here?

Reviewer: I suggested removing it from above, so it's not mentioned twice.

> Hierarchical configurations where one environment [imports](/docs/esc/environments/#importing-other-environments) others are fundamental to Pulumi ESC. If an environment is importing another environment, any changes made to the imported environment are also received as a Webhook configured within the current environment. The Webhook will notify you that the environment has changed, and you can call back to open the environment using the REST API, Pulumi Service Provider, Automation API, and SDK, and get the latest secrets and configuration.

Reviewer: "change ... are received" might imply that we're sending a diff when we're not. Maybe just "changes trigger a webhook"

> Hierarchical configurations where one environment [imports](/docs/esc/environments/#importing-other-environments) others are fundamental to Pulumi ESC. If an environment is importing another environment, any changes made to the imported environment are also received as a Webhook configured within the current environment. The Webhook will notify you that the environment has changed, and you can call back to open the environment using the REST API, Pulumi Service Provider, Automation API, and SDK, and get the latest secrets and configuration.

Author: ```suggestion
Hierarchical configurations where one environment [imports](/docs/esc/environments/#importing-other-environments) others are fundamental to Pulumi ESC. If an environment is importing another environment, any changes made to the imported environment trigger a Webhook configured within the current environment. The Webhook will notify you that the environment has changed, and you can call back to open the environment using the REST API, Pulumi Service Provider, Automation API, and SDK, and get the latest secrets and configuration.
```

## Getting Started with Pulumi ESC Webhooks

Setting up Webhooks is straightforward. You can use the Pulumi Cloud Console, the ESC SDK, the Pulumi Service Provider, or REST API.

### Using Pulumi Cloud Console

- Navigate to the "Webhooks" tab under settings for organization-wide Webhooks or within an individual environment's top navigation bar.

> - Select your desired integration (Slack, Microsoft Teams, or generic JSON Webhooks).

Reviewer: or Deployments!

> - Select your desired integration (Slack, Microsoft Teams, or generic JSON Webhooks).

Author: ```suggestion
- Select your desired integration (Slack, Microsoft Teams, Pulumi Deployments, or generic JSON Webhooks).
```

- Configure the Webhook details, including the events you want to trigger notifications on.

<b>(Add a video demo/image)</b>

### Using Pulumi Service Provider

> <b>(To be added)</b>

Reviewer: You can grab example from PR https://github.com/pulumi/pulumi-pulumiservice/pull/406

Something like:
```
const environmentWebhook = new service.Webhook("env-webhook", {
  active: true,
  displayName: "env-webhook",
  organizationName: "my-org",
  projectName: environment.project,
  environmentName: environment.name,
  payloadUrl: "https://example.com",
  filters: [WebhookFilters.EnvironmentRevisionCreated, WebhookFilters.ImportedEnvironmentChanged],
})
```

### Using the SDK

> <b>(To be added)</b>

Reviewer: This one https://github.com/pulumi/pulumi-service/issues/20923 was a could have and we probably won't have it

---
title: "Introducing Webhooks for Pulumi ESC"
date: 2024-09-13T00:00:00-03:00
draft: false
meta_desc: "Pulumi ESC Webhooks enable you to automate workflows, trigger actions, and streamline the management of your infrastructure and applications"
meta_image: "meta.png"
authors:
  - iaroslav-titov
  - arun-loganathan
tags:
  - esc
  - secrets
  - features
---

Managing secrets and configurations across multiple environments and teams can be a complex juggling act for development teams. [Pulumi ESC](/docs/esc/), our developer-friendly secrets management product, simplifies this process and accelerates your development cycle. Today, we're excited to make this process even easier with the launch of Webhooks for Pulumi ESC, a powerful new feature that allows you to send a custom trigger to any URL endpoint, enabling you to automate workflows and keep your infrastructure and applications up to date. Pulumi ESC Webhooks also integrates seamlessly with popular communication platforms like Slack and Microsoft Teams.

<!--more-->

> With Pulumi ESC Webhooks, you can now receive real-time notifications and trigger automated actions based on specific events within your Pulumi ESC [Environments](/docs/esc/environments/).  For example, you can set up a Webhook to notify your team’s Slack channel whenever a new secret is added to an environment, or automatically trigger a redeployment pipeline when configurations are changed. Webhooks can be triggered by changes to [version tags](/docs/esc/environments/#tagging-versions) and [environment tags](/blog/esc-projects-environment-tags-launch/#introducing-environment-tags), and can be configured per environment or at the organization level for all environments.

Reviewer: ```suggestion
With Pulumi ESC Webhooks, you can now receive real-time notifications and trigger automated actions based on specific events within your Pulumi ESC [Environments](/docs/esc/environments/).  For example, you can set up a Webhook to notify your team’s Slack channel whenever a revision tag was updated, or receive a POST request for each new environment created. Webhooks can be triggered by changes to [version tags](/docs/esc/environments/#tagging-versions) and [environment tags](/blog/esc-projects-environment-tags-launch/#introducing-environment-tags), and can be configured per environment or at the organization level for all environments.
```
